incubator/virtualcluster/pkg/syncer/controllers/pod/checker.go

```diff
@@ -0,0 +1,138 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+
+	ctrl "github.com/kubernetes-sigs/multi-tenancy/incubator/virtualcluster/pkg/syncer/controller"
+	"github.com/kubernetes-sigs/multi-tenancy/incubator/virtualcluster/pkg/syncer/conversion"
+)
+
+// StartPeriodChecker starts the period checker for data consistency check. Checker is
+// blocking so should be called via a goroutine.
+func (c *controller) StartPeriodChecker(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	// Start a loop to periodically check if pods keep consistency between super
+	// master and tenant masters.
+	wait.Until(c.checkPods, c.periodCheckerPeriod, stopCh)
+}
+
+// checkPods checks to see if pods in super master informer cache and tenant master
+// keep consistency.
+func (c *controller) checkPods() {
+	clusterNames := c.multiClusterPodController.GetClusterNames()
+	if len(clusterNames) == 0 {
+		klog.Infof("tenant masters has no clusters, give up period checker")
+		return
+	}
+
+	for _, clusterName := range clusterNames {
+		cluster := c.multiClusterPodController.GetCluster(clusterName)
+		if cluster == nil {
+			klog.Errorf("failed to locate cluster %s", clusterName)
+			continue
+		}
+
+		c.checkPodsOfCluster(clusterName, cluster)
+	}
+
+	pPods, err := c.podLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("error listing pods from super master informer cache: %v", err)
+		return
+	}
+
+	for _, pPod := range pPods {
+		clusterName, vNamespace := conversion.GetOwner(pPod)
+		vPodObj, err := c.multiClusterPodController.Get(clusterName, vNamespace, pPod.Name)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				if pPod.DeletionTimestamp != nil {
+					// pPod is under deletion, waiting for UWS bock populate the pod status.
+					continue
+				}
+				if err = c.client.Pods(pPod.Namespace).Delete(pPod.Name, &metav1.DeleteOptions{}); err != nil {
+					klog.Errorf("error deleting pPod %s/%s in super master: %v", pPod.Namespace, pPod.Name, err)
+					continue
+				}
+			}
+		}
+		vPod := vPodObj.(*v1.Pod)
+
+		// pod has been updated by super master
+		if !equality.Semantic.DeepEqual(vPod.Status, pPod.Status) {
+			c.enqueuePod(pPod)
+		}
+	}
+}
+
+// checkPodsOfCluster checks to see if pods in specific cluster keeps consistency.
+func (c *controller) checkPodsOfCluster(clusterName string, cluster ctrl.ClusterInterface) {
+	clusterInformerCache, err := cluster.GetCache()
+	if err != nil {
+		klog.Errorf("failed to get informer cache for cluster %s", clusterName)
+		return
+	}
+	podList := &v1.PodList{}
+	err = clusterInformerCache.List(context.TODO(), podList)
+	if err != nil {
+		klog.Errorf("error listing pods from cluster informer cache: %v", err)
+		return
+	}
+
+	klog.Infof("check pods consistency in cluster %s", clusterName)
+	for _, vPod := range podList.Items {
+		targetNamespace := conversion.ToSuperMasterNamespace(clusterName, vPod.Namespace)
+		pPod, err := c.podLister.Pods(targetNamespace).Get(vPod.Name)
+		if err == nil {
```

Comment:
should firstly check if vPod is under deletion. if vPod is under deletion, no matter what status pPod is, delete pPod.

else if vPod is exists and pPod not found, recreate one(or enqueue)

Reply:
most of the time, pPod should be deleted by kubelet, period checker is just used for offset to the lose.